import time
from telebot import TeleBot
import requests
from app.core.config import get_settings
from app.core.logger import setup_logger
from app.Text_messages.logs import LogMessages

logger = setup_logger()

RETRY_INTERVAL = 10 # ÿ´ÿßŸÜ€åŸá ÿ®€åŸÜ Ÿáÿ± ÿ™ŸÑÿßÿ¥ ŸÖÿ¨ÿØÿØ

settings = get_settings()

def create_bot_with_retry():
    

    while True:
        try:
            logger.info(LogMessages.STARTUP_TRY_CONNECT)
            bot = TeleBot(settings.BOT_TOKEN)
            bot.get_me()  # ÿ®ÿ±ÿ±ÿ≥€å ÿßÿ™ÿµÿßŸÑ ŸàÿßŸÇÿπ€å
            logger.info(LogMessages.STARTUP_SUCCESS)
            return bot

        except requests.exceptions.RequestException:
            logger.warning(LogMessages.UNEXPECTED_SHUTDOWN)
            time.sleep(RETRY_INTERVAL)

        except Exception as e:
            logger.exception(LogMessages.STARTUP_FAILED.format(error=e))
            time.sleep(RETRY_INTERVAL)


def run_bot_white_try(bot:TeleBot):
    
    while True:
        try:
            logger.info("üî• Bot starting...")
            bot.polling(non_stop=True)
        except requests.exceptions.ProxyError:
            logger.warning("‚ö†Ô∏è Proxy Error. Retrying in %s seconds...", RETRY_INTERVAL)
            time.sleep(RETRY_INTERVAL)
        except requests.exceptions.RequestException:
            logger.warning("‚ùå Unexpected shutdown. Retrying in %s seconds...", RETRY_INTERVAL)
            time.sleep(RETRY_INTERVAL)
        except requests.exceptions.ConnectionError:
            logger.warning("‚ùåConnectionError   %s seconds ", RETRY_INTERVAL)
            time.sleep(RETRY_INTERVAL)
        except requests.exceptions.Timeout:
            logger.warning("‚ùå Timeout   %s seconds ", RETRY_INTERVAL)
            time.sleep(RETRY_INTERVAL)    
        except Exception as e:
            logger.exception("üö® Unknown error: %s", e)
            time.sleep(RETRY_INTERVAL)